web3_proxy/src/frontend/users/payment.rs

```diff
@@ -413,60 +421,35 @@ pub async fn user_balance_post(
-                active_user_balance.available_balance = sea_orm::Set(balance_plus_amount);
-
-                if balance_plus_amount >= Decimal::new(10, 0) {
-                    // Also make the user premium at this point ...
-                    let mut active_recipient = recipient.clone().into_active_model();
-                    // Make the recipient premium "Effectively Unlimited"
-                    active_recipient.user_tier_id = sea_orm::Set(premium_user_tier.id);
-                    active_recipient.save(&txn).await?;
-                }
-
-                debug!("New user balance model is: {:?}", active_user_balance);
-                active_user_balance.save(&txn).await?;
-                // txn.commit().await?;
-                // user_balance
-            }
-            None => {
-                // Create the entry with the respective balance
-                let active_user_balance = balance::ActiveModel {
-                    available_balance: sea_orm::ActiveValue::Set(amount),
-                    user_id: sea_orm::ActiveValue::Set(recipient.id),
-                    ..Default::default()
-                };
-
-                if amount >= Decimal::new(10, 0) {
-                    // Also make the user premium at this point ...
-                    let mut active_recipient = recipient.clone().into_active_model();
-                    // Make the recipient premium "Effectively Unlimited"
-                    active_recipient.user_tier_id = sea_orm::Set(premium_user_tier.id);
-                    active_recipient.save(&txn).await?;
-                }
-
-                info!("New user balance model is: {:?}", active_user_balance);
-                active_user_balance.save(&txn).await?;
-                // txn.commit().await?;
-                // user_balance // .try_into_model().unwrap()
-            }
-        };
+        // Gotta add to the balance entry
+        // Should be atomic, because we do a write lock during the transaction
+        let mut active_user_balance = user_balance.clone().into_active_model();
+        active_user_balance.available_balance =
+            sea_orm::Set(user_balance.available_balance + amount);
+
+        // I can now call update / save and get the new object back
+        let mut user_balance = active_user_balance.save(&txn).await?;
+
+        // Then also make the user premium if the payment is above 10$
+        if user_balance.available_balance.unwrap() >= Decimal::new(10, 0) {
+            let mut active_recipient = recipient.clone().into_active_model();
+            // Make the recipient premium "Effectively Unlimited"
+            active_recipient.user_tier_id = sea_orm::Set(premium_user_tier.id);
+            active_recipient.save(&txn).await?;
+        }
+
         debug!("Setting tx_hash: {:?}", tx_hash);
         let receipt = increase_on_chain_balance_receipt::ActiveModel {
             tx_hash: sea_orm::ActiveValue::Set(hex::encode(tx_hash)),
```

Comment:
hex::encode will not work correctly. It does not include the 0x prefix. 

Use ether's types and their .encode_hex() functions instead. They already do the correct thing.